src/librustc_mir/build/matches/mod.rs

```diff
@@ -308,11 +309,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                     block,
                     Statement {
                         source_info: ty_source_info,
-                        kind: StatementKind::AscribeUserType(
-                            place,
-                            ty::Variance::Invariant,
-                            user_ty,
-                        ),
+                        kind: StatementKind::AscribeUserType(place, variance, user_ty),
```

Comment:
I think here we should write:

```rust
kind: StatementKind::AscribeUserType(place, ty::Invariant, user_ty),
```

and add a comment like this:

Note that we always use *invariant* as the variance here. This is because the `variance` field from the ascription refers to the variance to use when applying the type to the value being matched, but this ascription applies rather to the type of the binding. e.g., in this example:

```
let x: T = <expr>
```

we are creating an ascription that defines the type of `x` to be exactly `T` (i.e., with invariance). The `variance` field, in contrast, is intended to be used to relate `T` to the type of `<expr>`.

Reply:
Changed.